common/djangoapps/student/tests/tests.py

```diff
@@ -930,3 +935,231 @@ def test_roundtrip_with_unicode_course_id(self):
+
+    @ddt.data('verified', 'honor')
+    def test_modes_program_courses_on_dashboard_with_configuration(self, course_mode):
+        """Test that if program configuration is enabled than student can only
+        see those courses with xseries upsell messages which are active in
+        xseries programs.
+        """
+        CourseEnrollment.enroll(self.user, self.course_1.id, mode=course_mode)
+        CourseEnrollment.enroll(self.user, self.course_2.id, mode=course_mode)
+
+        self.client.login(username="jack", password="test")
+        self.create_config(enabled=True, enable_student_dashboard=True)
+
+        with patch('student.views.get_course_programs_for_dashboard') as mock_data:
+            mock_data.return_value = self._create_program_data(
+                [(self.course_1.id, 'active'), (self.course_2.id, 'unpublished')]
+            )
+            response = self.client.get(reverse('dashboard'))
+            # count total courses appearing on student dashboard
+            self.assertContains(response, 'course-container', 2)
+            self._assert_responses(response, 1)
+
+            # for verified enrollment view the program detail button will have
+            # the class 'base-btn'
+            # for other modes view the program detail button will have have the
+            # class border-btn
+            if course_mode == 'verified':
+                self.assertIn('xseries-base-btn', response.content)
+            else:
+                self.assertIn('xseries-border-btn', response.content)
+
+    @ddt.data(
+        ('unpublished', 'unpublished', 'unpublished', 0),
+        ('active', 'unpublished', 'unpublished', 1),
+        ('active', 'active', 'unpublished', 2),
+        ('active', 'active', 'active', 3),
+    )
+    @ddt.unpack
+    def test_different_programs_on_dashboard(self, status_1, status_2, status_3, program_count):
+        """Test the upsell on student dashboard with different programs
+        statuses.
+        """
+
+        CourseEnrollment.enroll(self.user, self.course_1.id, mode='verified')
+        CourseEnrollment.enroll(self.user, self.course_2.id, mode='honor')
+        CourseEnrollment.enroll(self.user, self.course_3.id, mode='honor')
+
+        self.client.login(username="jack", password="test")
+        self.create_config(enabled=True, enable_student_dashboard=True)
+
+        with patch('student.views.get_course_programs_for_dashboard') as mock_data:
+            mock_data.return_value = self._create_program_data(
+                [(self.course_1.id, status_1),
+                 (self.course_2.id, status_2),
+                 (self.course_3.id, status_3)]
+            )
+
+            response = self.client.get(reverse('dashboard'))
+            # count total courses appearing on student dashboard
+            self.assertContains(response, 'course-container', 3)
+            self._assert_responses(response, program_count)
+
+    @patch('student.views.log.warn')
```

Comment:
This is hiding that fact that `log.warn` is a non-existent method. When possible, you should test logging using `testfixtures.LogCapture`. Since `textfixtures` isn't currently an edx-platform depedency, I think it's okay to fix the method being mocked here.